### PR TITLE
New arg: `compute_with_cache`

### LINKED ISSUE #1738: Accuracy calculation blowing up memory use
### To Reproduce

Steps to reproduce the behavior...

<details>
  <summary>Code sample</summary>

```python
import torch
from torchmetrics import Accuracy

print(torch.cuda.memory_allocated()/1024**2)
# 0.0

device = "cuda"


x = torch.randn(8, 512, 30522, device=device)
preds = x.argmax(dim = -1)
y = torch.randint(0, 30522, (8, 512), device=device)
print(torch.cuda.memory_allocated()/1024**2)
# 476.96875

for i in range(100):
  gen_accuracy = Accuracy(task="multiclass", num_classes=30522, ignore_index=-100, average = 'micro', multidim_average = 'samplewise').to(device)
  acc1 = torch.sum((preds == y).float(), dim = -1) #doesn't blow up memory
  acc2 = gen_accuracy(preds, y) #does blow up memory
  print(torch.cuda.memory_allocated()/1024**2)
```

</details>

### Expected behavior

Memory footprint stays constant and does not increase

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source):
- Python & PyTorch Version (e.g., 1.0):
- Any other relevant information such as OS (e.g., Linux):

Running this on google colab with a `!pip install torchmetrics` right before should be sufficient

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1738 
Adds a new global arg `compute_with_cache`, that determines if the result from `compute` should be stored in a cache or not. By default it is `True` to be backwards compatible, but in some use cases (see the issue) it is beneficial to set this to `False` to essentially help the python garbage collector better clean up memory.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
I am in favor of not introduce this argument. 

Yes, we have this one caching issue with memory leakage due to python garbage collection (https://github.com/Lightning-AI/torchmetrics/issues/1738) . I am in favor of just removing the caching of output values and simply compute if `.compue` is called. This is extremely straight forward and actually the expected way IMO (I get that it's nice saving time by avoiding recomputes, but IMO this would be up to the user to do by caching the value themselves if they want to reuse it).

Other opinions?

cc @lantiga @Borda for visibility

**Comment by Borda** [comment]
cc: @lantiga ^^ :flamingo:

**Review by lantiga — APPROVED** [review]
Looks great, backwards compatible and adds opt-in control.

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
